Index: edu/cmu/pact/BehaviorRecorder/Tab/CTATTabManager.java
===================================================================
--- edu/cmu/pact/BehaviorRecorder/Tab/CTATTabManager.java	(revision 20342)
+++ edu/cmu/pact/BehaviorRecorder/Tab/CTATTabManager.java	(working copy)
@@ -5,21 +5,18 @@
 import java.io.File;
 import java.net.Socket;
 import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
-import javax.swing.JMenuItem;
 import javax.swing.JOptionPane;
 
 import net.infonode.docking.OperationAbortedException;
+
 import edu.cmu.pact.BehaviorRecorder.Controller.BR_Controller;
 import edu.cmu.pact.BehaviorRecorder.Controller.CTAT_Launcher;
 import edu.cmu.pact.BehaviorRecorder.Controller.SingleSessionLauncher;
 import edu.cmu.pact.SocketProxy.SocketProxy;
 import edu.cmu.pact.Utilities.trace;
-import edu.cmu.pact.ctatview.CtatMenuBar;
 import edu.cmu.pact.ctatview.DockManager;
 
 /**
@@ -43,12 +40,7 @@
 	private static int MAX_TABS = 5;
 	/** The number of graph tabs created. Incremented by {@link #getNewTab()}. */
 	private static int NUM_TABS = 0;
-
-	/** Master set of menu items defined on {@link CtatMenuBar}, for initializing tabs. */
-	private HashMap<JMenuItem, boolean[]> menuItems = new HashMap<JMenuItem, boolean[]>();
 	
-	/** Set of menu item names to de*/
-	
 	public CTATTabManager(CTAT_Launcher server, String[] argv) {
 		this.argv = argv;
 		this.tabList = new ArrayList<CTATTab>();
@@ -57,23 +49,6 @@
 	}
 	
 	/**
-	 * Add a menu item to {@link #menuItems} and call
-	 * {@link CTATTab#addCtatMenuItem(CtatMenuBar.MenuItem, boolean[])} for each tab.
-	 * Call this from the menu item's constructor to record the new instance in the sets
-	 * of menu items to save.
-	 * @param menuItem
-	 * @param enabled true if the menu item is initially enabled; false if disabled
-	 */
-	public void addCtatMenuItem(JMenuItem menuItem, boolean[] values) {
-		if(trace.getDebugCode("menu"))
-			trace.out("menu", "CTATTabManager.addCtatMenuItems("+menuItem+") nInField "+
-					menuItems.size()+", tabList size "+tabList.size());
-		menuItems.put(menuItem, values);
-		for(CTATTab tab : tabList)
-			tab.addCtatMenuItem(menuItem, Arrays.copyOf(values, values.length));
-	}
-	
-	/**
 	 * @return the {@link #nextTabNumber}
 	 */
 	public int getNextTabNumber() {
@@ -139,7 +114,7 @@
 	 * 					{@link BR_Controller}.
 	 */
 	public CTATTab getNewTab(CTATTab newTab, SingleSessionLauncher launcher) {
-		if (NUM_TABS >= MAX_TABS) {
+        if (NUM_TABS >= MAX_TABS) {
 			if(trace.getDebugCode("mg"))
 				trace.out("mg", "CTATTabManager (getNewTab): no more tabs ");
 			newTab = this.getFreeTab();
@@ -158,7 +133,6 @@
 		this.tabList.add(newTab);
 		NUM_TABS++;
 		this.nextTabNumber++;
-		newTab.initializeCtatMenuItems(menuItems);
 		return newTab;
 	}
 	
@@ -216,7 +190,6 @@
 		DockManager dockManager = this.server.getDockManager();
 		
 		this.focusedTab = tab;
-		this.focusedTab.updateCtatMenuItems();
 		if(!this.server.isDoneIntializing()) return;
 		dockManager.markAsFocused(tab.getTabNumber(), lastFocusedTabNumber);
 		if(reloadViews && dockManager != null) {
@@ -360,7 +333,7 @@
 			} else { // if Java
 				// either close the graph *and* interface, or
 				// clear the graph and abort the close, leaving the interface connected.
-				if (ctlr.getStudentInterface() != null) {
+                if (ctlr.getStudentInterface() != null) {
 					int reply = JOptionPane.showConfirmDialog(ctlr.getActiveWindow(),
 							"Closing the window "+tab.getName()+" will also close the connected interface. \r\n" +
 							"Cancel to leave the interface open and connected.  Continue?",
Index: edu/cmu/pact/BehaviorRecorder/View/HintWindow/HintPanel.java
===================================================================
--- edu/cmu/pact/BehaviorRecorder/View/HintWindow/HintPanel.java	(revision 20342)
+++ edu/cmu/pact/BehaviorRecorder/View/HintWindow/HintPanel.java	(working copy)
@@ -18,6 +18,7 @@
 import javax.swing.JButton;
 import javax.swing.JEditorPane;
 import javax.swing.JLabel;
+import javax.swing.JOptionPane;
 import javax.swing.JPanel;
 import javax.swing.JScrollPane;
 
Index: edu/cmu/pact/client/HintMessagesManagerForClient.java
===================================================================
--- edu/cmu/pact/client/HintMessagesManagerForClient.java	(revision 20342)
+++ edu/cmu/pact/client/HintMessagesManagerForClient.java	(working copy)
@@ -24,6 +24,7 @@
 import java.util.regex.Pattern;
 
 import javax.swing.BorderFactory;
+import javax.swing.JOptionPane;
 import javax.swing.border.Border;
 
 import pact.CommWidgets.JCommButton;
@@ -78,6 +79,7 @@
     	private final Border border;
     	private final String action;
     	public HighlightWidget(String selection, Border border, String action) {
+    		
         	this.selection = selection;
         	this.border = border;
         	this.action = action;
@@ -158,6 +160,7 @@
 	 * @see edu.cmu.pact.client.HintMessagesManager#setMessageObject(edu.cmu.pact.ctat.MessageObject)
 	 */
     public void setMessageObject(MessageObject o) {
+    
         reset();
         hintsMessageObject = o;
         parseMessages();
@@ -409,6 +412,7 @@
     }
 
     protected JCommWidget selectedWidget() {
+    	
         return controller==null ? null : controller.getCommWidget(selectionName);
     }
     
@@ -631,6 +635,9 @@
     }
 
     protected Vector getHintMessageVector(String propertyName) {
+    	
+    
+    	
  //       return (Vector)getHintMessageProperty(propertyName);
     	Object result = getHintMessageProperty(propertyName);
     	
@@ -754,6 +761,8 @@
     private void processShowHintsMessage() {
     	stepID = (String) getHintMessageProperty("StepID");
 
+    	
+    	
         Vector selectionNamesVector = getHintMessageSelection();
         Vector actionVector = getHintMessageAction();
 
@@ -761,7 +770,7 @@
             highlightFlag = false;
             return;
         }
-
+     
         trace.out(5, this, "selectionPart.size() = " + selectionNamesVector.size());
 
         if (selectionNamesVector.size()==0) {
Index: edu/cmu/pact/ctat/TutorController.java
===================================================================
--- edu/cmu/pact/ctat/TutorController.java	(revision 20342)
+++ edu/cmu/pact/ctat/TutorController.java	(working copy)
@@ -841,6 +841,11 @@
     	// zz add 11/06/03: deal with the same commName IgnoreCase
 
     	// chc added 12/13/06 to allow Author create new Blank state with "No_Selection".
+    	
+
+    	
+    	
+    	
     	if (commName.equals("No_Selection")) return null;
 
     	Enumeration<String> iterKeys = getCommWidgetTable().keys();
Index: edu/cmu/pact/ctatview/CtatMenuBar.java
===================================================================
--- edu/cmu/pact/ctatview/CtatMenuBar.java	(revision 20342)
+++ edu/cmu/pact/ctatview/CtatMenuBar.java	(working copy)
@@ -164,7 +164,7 @@
 		/** Add a new item to the saved set. Call only from constructors. */
 		private void storeInTabs() {
 			boolean[] values = makeValues();
-			server.getTabManager().addCtatMenuItem(this, values);
+			//server.getTabManager().addCtatMenuItem(this, values);
 		}
 		
 		/** Create an array to store the values of interest. */
@@ -280,7 +280,7 @@
 		/** Add a new item to the saved set. Call only from constructors. */
 		private void storeInTabs() {
 			boolean[] values = makeValues();
-			server.getTabManager().addCtatMenuItem(this, values);
+			//server.getTabManager().addCtatMenuItem(this, values);
 		}
 		
 		/** Create an array to store the values of interest. */
Index: edu/cmu/pact/jess/IsHint.java
===================================================================
--- edu/cmu/pact/jess/IsHint.java	(revision 20342)
+++ edu/cmu/pact/jess/IsHint.java	(working copy)
@@ -65,8 +65,8 @@
 			return null;
 		if (context.getEngine() instanceof MTRete)
 			return ((MTRete) context.getEngine()).getJessModelTracing();
-		if(context.getEngine() instanceof SimStRete) 
-			return ((SimStRete)context.getEngine()).getJmt();
+		//if(context.getEngine() instanceof SimStRete) 
+			//return ((SimStRete)context.getEngine()).getJmt();
 		return null;
 	}
 }
Index: edu/cmu/pact/jess/JessModelTracing.java
===================================================================
--- edu/cmu/pact/jess/JessModelTracing.java	(revision 20342)
+++ edu/cmu/pact/jess/JessModelTracing.java	(working copy)
@@ -437,6 +437,8 @@
 		
 		wmImages = null;
 		
+
+		
 		//System.out.println("entered JessModelTracing's modelTrace(...)   REMOVE THIS PRINT");
 
 		if (trace.getDebugCode("mt")) trace.out("mt", "modelTrace() rete="+rete.hashCode());
Index: edu/cmu/pact/jess/MT.java
===================================================================
--- edu/cmu/pact/jess/MT.java	(revision 20342)
+++ edu/cmu/pact/jess/MT.java	(working copy)
@@ -8,6 +8,7 @@
 import java.io.Reader;
 import java.net.URL;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.Enumeration;
@@ -18,7 +19,6 @@
 import java.util.StringTokenizer;
 import java.util.Vector;
 
-
 import jess.Deftemplate;
 import jess.Fact;
 import jess.FactIDValue;
@@ -314,7 +314,7 @@
 	 * @param rete
 	 * @param jmt
 	 */
-    public static void loadDefaultUserfunctions(Object obj, Rete rete, JessModelTracing jmt) {
+    public static void loadDefaultUserfunctions(MT obj, Rete rete, JessModelTracing jmt) {
 		String thisPkg = obj.getClass().getPackage().getName();
 		String[] userfunctions = {
 			thisPkg+".ConstructMessage",	
@@ -340,6 +340,7 @@
 			thisPkg+".SetHintPolicy",
 			thisPkg+".SetDefaultSkillCategory"
 		};
+		
 		FunctionPackage fp = new FunctionPackage(userfunctions, jmt);
 		if (fp.size() < userfunctions.length)
 			trace.err("Jess Model Tracer Warning: only "+fp.size()+" of "+
@@ -382,28 +383,8 @@
      */
 	MT(CTAT_Controller controller, MTRete externallyDefinedRete){
 		if (trace.getDebugCode("mt")) trace.out("mt", "MT constructor");
-		this.externallyDefinedRete = externallyDefinedRete;
 		this.controller = controller;
-
-		clearRete();  // call early this to set rete for logging
-
-		// rete = new MTRete(controller);
-		// rete.setMT(this);
-		if(VersionInformation.includesSimSt() && !Utils.isRuntime())
-			jmt = new SimStJessModelTracing(rete, controller);
-		else
-			jmt = new JessModelTracing(rete, controller);
-		rete.setJmt(jmt);
-		jmt.setRete(rete);
-		
-
-		if(!Utils.isRuntime()) {
-			wmeEditor = new WMEEditor(rete, this, null, true, false);
-			wmeEditor.setProblemName(controller.getLogger().getStudentName());
-			addMessageEventListener(wmeEditor);
-			
-			console = new JessConsole("Jess Console", rete, true, controller);
-		}		
+		this.setExternalRete(externallyDefinedRete);
     }
     
     /**
@@ -600,6 +581,14 @@
     
     /**
      * Get the directory named by the preference {@value CTAT_Controller#PROBLEM_DIRECTORY}.
+     * @return preference value, with trailing {@link File#separator}
+     */
+    String getCognitiveModelDirectory() {
+    	return getPreferenceDirectory(CTAT_Controller.PROBLEM_DIRECTORY, ".", null);
+    }
+    
+    /**
+     * Get the directory named by the preference {@value CTAT_Controller#PROBLEM_DIRECTORY}.
      * @param subdir optional subdirectory to append
      * @return preference value, with trailing {@link File#separator}
      */
@@ -705,16 +694,13 @@
 //            System.gc();
             MTRete.stopModelTracing = false;
             String fullName = getController().getProblemFullName();
-            if(fullName != null && fullName.length() > 0) {
+            if(fullName != null) {
             	String[] path = fullName.split("[\\\\/]");
             	String baseName = path[path.length-1];
             	if(baseName.toLowerCase().endsWith(".brd"))
             		problemName = baseName.substring(0, baseName.length()-4);
             	else
             		problemName = baseName;
-            } else {
-            	Object name = mo.getProperty(MsgType.PROBLEM_NAME);
-            	problemName = (name instanceof String ? name.toString() : "");
             }
             return null;
         }else if (messageType.equalsIgnoreCase("StartStateEnd")) {
@@ -747,16 +733,20 @@
                     this.rete.eval("(defglobal ?*sAction* = "+MTRete.NOT_SPECIFIED+")");
                     this.rete.eval("(defglobal ?*sInput* = "+MTRete.NOT_SPECIFIED+")");
 
+                    /*
                     // define a deftemplate for storing the student input values.
-//                    if(rete.findDeftemplate("studentValues") == null){
-//                        String deftemplate = "(deftemplate studentValues (slot selection) (slot action) (slot input))";
-//                        rete.eval(deftemplate);
-//                    }
+                    if(rete.findDeftemplate("studentValues") == null){
+                        String deftemplate = "(deftemplate studentValues (slot selection) (slot action) (slot input))";
+                        rete.eval(deftemplate);
+                    }
+                    */
                 }
 
+                /*
                 // loading the rules from the file "productionRules.pr"
-//              if (!prRulesFromFile)
-//              prRulesFromFile = pre.loadRuleSet();
+                if (!prRulesFromFile)
+                prRulesFromFile = pre.loadRuleSet();
+				*/
 
             } catch (HeadlessException e1) {
                 e1.printStackTrace();
@@ -1569,6 +1559,29 @@
      */
     private Vector startStateHooks;
 	
+    public void setExternalRete(MTRete external){
+    	this.externallyDefinedRete = external;
+
+		clearRete();  // call early this to set rete for logging
+
+		// rete = new MTRete(controller);
+		// rete.setMT(this);
+		if(VersionInformation.includesSimSt() && !Utils.isRuntime())
+			jmt = new SimStJessModelTracing(rete, controller);
+		else
+			jmt = new JessModelTracing(rete, controller);
+		rete.setJmt(jmt);
+		jmt.setRete(rete);
+
+		if(!Utils.isRuntime()) {
+			wmeEditor = new WMEEditor(rete, this, null, true, false);
+			wmeEditor.setProblemName(controller.getLogger().getStudentName());
+			addMessageEventListener(wmeEditor);
+			
+			console = new JessConsole("Jess Console", rete, true, controller);
+		}		
+    }
+    
 	/**
 	 * Load the deftemplates and facts files for a given test.  If useBinary
 	 * is set, tries to Rete.bload() that file first.  Sets wmeTypesFromFile,
@@ -1581,14 +1594,12 @@
 	 * @return array showing which files were read
 	 */
 	private void loadJessFiles(String problemName, boolean useBinary) {
-            
+		
 		bloadSuccessful = false;
 		wmeTypesFromFile = false;
 		prRulesFromFile = false;
 		wmeInstancesFromFile = false;
 		
-		//String wmeInstanceFilename =problemName + ".wme";
-			
 		String[] filenames = { problemName + ".bload",             // cf use of
 							   WMEEditor.wmeTypeFileName,          // indices in
 							   WMEEditor.rulesFileName, // results[]
@@ -2143,4 +2154,9 @@
 		if(wmeEditor != null && wmeEditor.getPanel() != null)
 			wmeEditor.getPanel().refresh();
 	}
+
+	public JessModelTracing getJMT() {
+		// TODO Auto-generated method stub
+		return jmt;
+	}
 }
Index: edu/cmu/pact/jess/MTRete.java
===================================================================
--- edu/cmu/pact/jess/MTRete.java	(revision 20342)
+++ edu/cmu/pact/jess/MTRete.java	(working copy)
@@ -781,7 +781,7 @@
 	/**
 	 * Override to also clear {@link #buggyRules}, restore listener.
 	 */
-	public synchronized void clear() throws JessException {
+	public void clear() throws JessException {
 		super.clear();
 		setStrategyByName(getLastAuthorChosenStrategyName());
 //		buggyRules.clear();
@@ -1225,6 +1225,8 @@
 	boolean[] loadJessFiles(String bloadName, String templatesName, String rulesName,
 	         String problemFactsName, List interfaceTemplatesList) 
 	throws JessException {
+		
+		System.out.println("WRONG!!!!!!!!!!!!!!!!!!!!!!!!");
 
         resetLoadInterfacetemplatesFailed();
         HintFact.setHintFact(false, this);         // initialize before reading rules   
Index: edu/cmu/pact/jess/PredictObservableAction.java
===================================================================
--- edu/cmu/pact/jess/PredictObservableAction.java	(revision 20342)
+++ edu/cmu/pact/jess/PredictObservableAction.java	(working copy)
@@ -187,10 +187,7 @@
 	protected JessModelTracing getJmt() {
 		if (jmt == null) {
 			if (context != null) {
-				if (context.getEngine() instanceof MTRete)
-					jmt = ((MTRete) context.getEngine()).getJessModelTracing();
-				if(context.getEngine() instanceof SimStRete) 
-					jmt = ((SimStRete)context.getEngine()).getJmt();
+                jmt = ((MTRete) context.getEngine()).getJessModelTracing();
 			}
 		}
 		return jmt;
Index: edu/cmu/pact/jess/RuleActivationNode.java
===================================================================
--- edu/cmu/pact/jess/RuleActivationNode.java	(revision 20342)
+++ edu/cmu/pact/jess/RuleActivationNode.java	(working copy)
@@ -481,7 +481,7 @@
 			String predictedSelection, String predictedAction,
 			String predictedInput, String predictedInputTestFunction,
 			boolean isHint, Context ctx) {
-		
+
 		if(trace.getDebugCode("rr"))
 			trace.out("***isSAIFound() s " + selection + "=?" +
 					  predictedSelection + ", a " + action + "=?" +
@@ -492,7 +492,8 @@
 		final int I = 0x4;   // whether we've matched student's input
 		int result = 0;      // match results so far
 		int m;
-
+	
+		
 		m = testSingleValue(predictedSelection, selection, "selection",
 				MTRete.NOT_SPECIFIED, isHint, ctx);
 		if (m == NO_MATCH) {             // sewall 2013/11/01: special return for hint requests
@@ -942,6 +943,8 @@
 	    		msg = msg.trim();
 	    	if (msg.length() > 0)
 	    		tgtMsgList.add(msg);
+	    	
+	    		
 	    }
 	}
 	
Index: edu/cmu/pact/jess/SetDefaultSkillCategory.java
===================================================================
--- edu/cmu/pact/jess/SetDefaultSkillCategory.java	(revision 20342)
+++ edu/cmu/pact/jess/SetDefaultSkillCategory.java	(working copy)
@@ -60,8 +60,8 @@
 			return null;
 		if (context.getEngine() instanceof MTRete)
 			return ((MTRete) context.getEngine()).getJessModelTracing();
-		if(context.getEngine() instanceof SimStRete) 
-			return ((SimStRete)context.getEngine()).getJmt();
+		//if(context.getEngine() instanceof SimStRete) 
+			//return ((SimStRete)context.getEngine()).getJmt();
 		return null;
 	}
 }
Index: edu/cmu/pact/jess/SimStRete.java
===================================================================
--- edu/cmu/pact/jess/SimStRete.java	(revision 20342)
+++ edu/cmu/pact/jess/SimStRete.java	(working copy)
@@ -17,6 +17,7 @@
 import java.net.URL;
 import java.rmi.server.UID;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
@@ -25,7 +26,6 @@
 import javax.swing.JOptionPane;
 
 import sun.security.action.GetBooleanAction;
-
 import jess.Activation;
 import jess.Context;
 import jess.Defrule;
@@ -162,27 +162,27 @@
 	}
 
 	/** JessModelTracing */
-	private JessModelTracing jmt;
+	//private JessModelTracing jmt;
 	
-	public JessModelTracing getJmt() {
-		return jmt;
-	}
+//	public JessModelTracing getJmt() {
+//		return jmt;
+//	}
+//
+//	public void setJmt(JessModelTracing jmt) {
+//		this.jmt = jmt;
+//	}
 
-	public void setJmt(JessModelTracing jmt) {
-		this.jmt = jmt;
-	}
-
 	/** ModelTracer */
-	private transient MT mt;
+	//private transient MT mt;
 	
-	public MT getMt() {
-		return mt;
-	}
+//	public MT getMt() {
+//		return mt;
+//	}
+//
+//	private void setMt(MT mt) {
+//		this.mt = mt;
+//	}
 
-	private void setMt(MT mt) {
-		this.mt = mt;
-	}
-
 	/**	Console like text output facility */
 	private transient TextOutput textOutput = TextOutput.getNullOutput();
 	
@@ -202,8 +202,9 @@
 	public SimStRete(BR_Controller brController) {
 		super(brController);
 		controller = brController;
-		mt = new MT(brController);  // why not pass self as externally defined Rete?
-		jmt = mt.getModelTracing();
+//		mt = new MT(brController, this);  // why not pass self as externally defined Rete?
+//		jmt = mt.getModelTracing();
+		MT.loadDefaultUserfunctions(controller.getModelTracer(), this, null);
 		ConflictResolutionStrategy.setStrategy(this);
 	}
 	
@@ -213,7 +214,7 @@
 	 * @throws Exception 
 	 */
 	public void init(String problemName, boolean loadJessFiles) throws Exception {
-
+		
 		ArrayList<String> startElements = null;
 		if((getController().getMissController() != null) && (getController().getMissController().getSimStPLE() != null)){
 			
@@ -259,7 +260,6 @@
 	    	}
 		}
 		else {
-			
 			throw new Exception("Need to initialize the working memory with start state elements.");
 		}
 
@@ -267,7 +267,7 @@
 		String[] problem = SimStRete.PROBLEM_NAME.split("=");
 		//getController().resetAllWidgets();
 		if(loadJessFiles) {
-			MT.loadDefaultUserfunctions(getMt(), this, null);
+			MT.loadDefaultUserfunctions(controller.getModelTracer(), this, null);
 			initWMUsingJessFiles(problemName, false);
 		}
 		if(problem.length == 2 && !(problem[0].isEmpty()) && !(problem[1]).isEmpty()) {
@@ -493,9 +493,9 @@
     
 	/**	Clears the rule activation tree */
 	private void clearRuleActivationTree(){
-		if(jmt == null)
+		if(controller.getModelTracer().getJMT() == null)
 			return;
-		RuleActivationTree tree= jmt.getRuleActivationTree();
+		RuleActivationTree tree= controller.getModelTracer().getJMT().getRuleActivationTree();
 		if(tree != null){
 			tree.clearTree(null);
 		}
@@ -739,6 +739,8 @@
 				WMEEditor.rulesFileName, // results[]
 				"init.wme", problemName + ".wme" }; // below
 
+		System.out.println("ss" + Arrays.toString(filenames));
+
 		boolean[] results = null;
 		for (int i = 0; i < filenames.length; ++i) {
 			filenames[i] = getDirectory(filenames[i]) + // assume dir already
@@ -842,6 +844,21 @@
 			return result;
 	}
 
+	boolean[] loadJessFiles(String bloadName, String templatesName, String rulesName,
+	         String problemFactsName, List interfaceTemplatesList) 
+	throws JessException {
+		boolean[] temp = loadJessFiles(bloadName, templatesName, rulesName, "init.wme", problemFactsName, interfaceTemplatesList);
+		boolean[] output = new boolean[4];
+		
+		output[0] = temp[0];
+		output[1] = temp[1];
+		output[2] = temp[2];
+		output[3] = temp[4];
+		
+		return output;
+		
+	}
+
 	/**
 	 * Load the given deftemplates, facts and rules files. Any null file
 	 * argument will be skipped.
@@ -859,7 +876,7 @@
 	boolean[] loadJessFiles(String bloadName, String templatesName, String rulesName,
 	        String initFactsName, String problemFactsName, List interfaceTemplatesList)                
 	throws JessException {
-        
+		
 	    String[] filenames = {bloadName, templatesName,	rulesName, initFactsName, problemFactsName};
         final int templateFileIndex = 1;
 	    final int rulesFileIndex = 2;
Index: edu/cmu/pact/miss/AskHintInBuiltClAlgebraTutor.java
===================================================================
--- edu/cmu/pact/miss/AskHintInBuiltClAlgebraTutor.java	(revision 20342)
+++ edu/cmu/pact/miss/AskHintInBuiltClAlgebraTutor.java	(working copy)
@@ -9,6 +9,7 @@
 
 public class AskHintInBuiltClAlgebraTutor extends AskHintClAlgebraTutor {
 
+
     public static final String COMM_STEM = "commTable";
 
 	public AskHintInBuiltClAlgebraTutor(BR_Controller brController,
Index: edu/cmu/pact/miss/FeaturePredicate.java
===================================================================
--- edu/cmu/pact/miss/FeaturePredicate.java	(revision 20342)
+++ edu/cmu/pact/miss/FeaturePredicate.java	(working copy)
@@ -217,13 +217,14 @@
     public static Integer valueTypeForAlgebra( String value ) {
         int valueType = -1;
         if (EqFeaturePredicate.isValidSimpleSkill(value) || 
-                EqFeaturePredicate.isSkillOperand(value) ) {
+               EqFeaturePredicate.isSkillOperand(value) ) {
             valueType = TYPE_SKILL_OPERAND;
         } else if (EqFeaturePredicate.isArithmeticExpression(value)) {
             valueType = TYPE_ARITH_EXP;
         } else if (EqFeaturePredicate.isExprList(value)) {
             valueType = TYPE_EXP_LIST;
         }
+        // should this return something? 
         return new Integer(valueType);
     }
     
Index: edu/cmu/pact/miss/InquiryClSolverTutor.java
===================================================================
--- edu/cmu/pact/miss/InquiryClSolverTutor.java	(revision 20342)
+++ edu/cmu/pact/miss/InquiryClSolverTutor.java	(working copy)
@@ -248,7 +248,8 @@
 		private boolean isBasicSkill(String action) {
 			boolean isBasicSkill = 
 				"add".equals(action) || "subtract".equals(action) ||
-				"multiply".equals(action) || "divide".equals(action);
+				"multiply".equals(action) || "divide".equals(action) ||
+				"glarb".equals(action);
 			return isBasicSkill;
 		}
 
@@ -373,11 +374,16 @@
 				String clAction = null;
 				String clInput = null;
 				
-				String table = selection.substring(COMM_STEM.length(), COMM_STEM.length()+1);
-				int tableVal = Integer.parseInt(table);
-				String column = selection.substring(COMM_STEM.length()+3, COMM_STEM.length()+4);
+                //if(trace.getDebugCode("miss"))
+                //        trace.out("miss", selection);
+
+                selection.indexOf("_");
+				//String table = selection.substring(selection.indexOf("_")+1, selection.indexOf("_")+2);
+				//String table = selection.substring(COMM_STEM.length(), COMM_STEM.length()+1);
+				//int tableVal = Integer.parseInt(table);
+				String column = selection.substring(selection.indexOf("_C")+2, selection.indexOf("R"));
 				int colVal = Integer.parseInt(column);
-				if(tableVal == 3 || colVal == 3){
+				if(colVal == 3){
 					String[] tokens = input.split(" ");
 					if(tokens[0].toLowerCase().startsWith("combine")) {
 						clAction = "clt";
@@ -389,10 +395,12 @@
 					sai = new SAI("", clAction, clInput);
 				}
 				else {
-				    clAction = (tableVal == 1 && colVal == 1 ? "left" : "right");
+				    clAction = (colVal == 1 ? "left" : "right");
 				    clInput = input;	
 					sai = new SAI("", clAction, clInput);
 				}
+				//Chris suggestion: 
+				//sai = new SAI(selection, action , input);
 			}
 			String clAction = sai.getAction();	
 				
@@ -439,11 +447,11 @@
 				String clAction = null;
 				String clInput = null;
 				
-				String table = selection.substring(COMM_STEM.length(), COMM_STEM.length()+1);
-				int tableVal = Integer.parseInt(table);
-				String column = selection.substring(COMM_STEM.length()+3, COMM_STEM.length()+4);
+				//String table = selection.substring(COMM_STEM.length(), COMM_STEM.length()+1);
+				//int tableVal = Integer.parseInt(table);
+				String column = selection.substring(selection.indexOf("_C")+2, selection.indexOf("R"));
 				int colVal = Integer.parseInt(column);
-				if(tableVal == 3 || colVal == 3) {
+				if(colVal == 3) {
 					String[] tokens = input.split(" ");
 					if(tokens[0].toLowerCase().startsWith("combine")){
 						clAction = "clt";
@@ -454,10 +462,12 @@
 					}
 					sai = new SAI("", clAction, clInput);
 				} else {
-					clAction = (tableVal == 1 && colVal == 1 ? "left" : "right");
+					clAction = (colVal == 1 ? "left" : "right");
 					clInput = input;
 					sai = new SAI("", clAction, clInput);
 				}
+				//Chris suggestion
+				//sai = new SAI(selection, action, input);
 			}
 			
 			try {
@@ -488,33 +498,43 @@
 		// Ask for a hint on what to do next
 		private void getNextStep() {
 			SAI nextStep = whatToDoNext();
-			String msg = HINTMESSAGE + ";" + 
-				nextStep.getSelection() + ";" + nextStep.getAction() + ";" + nextStep.getInput();
+
+			if(trace.getDebugCode("ss"))
+				trace.out("ss", "Selection: " + nextStep.getSelection());
+				trace.out("ss", "Action: " + nextStep.getAction());
+				trace.out("ss", "Input: " + nextStep.getInput());
+
+			String msg = HINTMESSAGE + ";" + nextStep.getSelection() + ";" + nextStep.getAction() + ";" + nextStep.getInput();
+
 			if(trace.getDebugCode("rr"))
 				trace.out("rr", "getNextStep >> " + msg);
+
 			handleMessage(msg);
 		}
 
 		private SAI whatToDoNext() {
 			SAI nextStep = null;
-			if (isNonBasicSkillApplied()) {
-				for (int i = 0; i < 2; i++) {
-					if (typeinExpressions[i] != null) {
-						// i = 0 for "left" and 1 for "right"
-						nextStep = new SAI("", i == 0 ? "left" : "right", typeinExpressions[i]);
-						break;
-					}
-				}
-			} else {
-				nextStep = this.solverTutor.getNextStep("tutor");
-				if (nextStep != null && nextStep.getSelection() == null)
-					nextStep.setSelection(this.solverTutor.getCurrentProblem().toString());
+			//chris had this deleted
+			if (isNonBasicSkillApplied()) {			
+								for (int i = 0; i < 2; i++) {			
+									if (typeinExpressions[i] != null) {					
+										// i = 0 for "left" and 1 for "right"			
+										nextStep = new SAI("", i == 0 ? "left" : "right", typeinExpressions[i]);				
+										break;
+									}
+							}
+						} else {
+            nextStep = this.solverTutor.getNextStep("tutor");
 			}
 			
-			return nextStep;
+			
+            if (nextStep != null && nextStep.getSelection() == null){
+            	nextStep.setSelection(this.solverTutor.getCurrentProblem().toString());
+            }
+			
+            return nextStep;
 		}
 
-		private static final String COMM_STEM= "dorminTable";
 		// - * - * - * - * - * - * - * - * - * - * - * - * - * - * - * - * - * - * 
 		// - * - * - * - * - * - * - * - * - * - * - * - * - * - * - * - * - * - * 
 		// Ask the TutorSolver if the give SAI is correct 
Index: edu/cmu/pact/miss/SimSt.java
===================================================================
--- edu/cmu/pact/miss/SimSt.java	(revision 20342)
+++ edu/cmu/pact/miss/SimSt.java	(working copy)
@@ -11,7 +11,6 @@
 package edu.cmu.pact.miss;
 
 import java.awt.event.ActionListener;
-import java.awt.event.FocusEvent;
 import java.io.BufferedReader;
 import java.io.File;
 import java.io.FileInputStream;
@@ -66,9 +65,8 @@
 import jess.Value;
 import mylib.Combinations;
 import pact.CommWidgets.JCommTable;
+import pact.CommWidgets.JCommTable.TableExpressionCell;
 import pact.CommWidgets.StudentInterfaceWrapper;
-import pact.CommWidgets.JCommTable.TableCell;
-import pact.CommWidgets.JCommTable.TableExpressionCell;
 import aima.search.framework.Node;
 import aima.search.framework.Problem;
 import aima.search.framework.Search;
@@ -97,8 +95,8 @@
 import edu.cmu.pact.jess.MTRete;
 import edu.cmu.pact.jess.RuleActivationNode;
 import edu.cmu.pact.jess.RuleActivationTree;
+import edu.cmu.pact.jess.RuleActivationTree.TreeTableModel;
 import edu.cmu.pact.jess.SimStRete;
-import edu.cmu.pact.jess.RuleActivationTree.TreeTableModel;
 import edu.cmu.pact.miss.InquiryClAlgebraTutor.TutorServerTimeoutException;
 import edu.cmu.pact.miss.MetaTutor.APlusQuizProblemAbstractor;
 import edu.cmu.pact.miss.PeerLearning.SimStConversation;
@@ -114,7 +112,6 @@
 import edu.cmu.pact.miss.ProblemModel.Graph.SimStProblemGraph;
 import edu.cmu.pact.miss.console.controller.MissController;
 import edu.cmu.pact.miss.jess.ModelTraceWorkingMemory;
-import edu.cmu.pact.miss.jess.WorkingMemoryConstants;
 import edu.cmu.pact.miss.storage.FileZipper;
 import edu.cmu.pact.miss.userDef.algebra.EqFeaturePredicate;
 
@@ -565,7 +562,7 @@
     }
     
     //11Dec2007: this flag is only used for the purposes of making a video.
-    private boolean useCacheOracleInquiry = true;
+    private boolean useCacheOracleInquiry = false;
     public boolean isUseCacheOracleInquiry(){ return this.useCacheOracleInquiry; }
     public void setSsCacheOracleInquiry(boolean flag) {
         this.useCacheOracleInquiry = flag;
@@ -2953,6 +2950,7 @@
         // Set Miss Controller
         setMissController( missController );
         setSsRete(new SimStRete(getBrController()));
+        getBrController().getModelTracer().setExternalRete(this.ssRete);
         
         logger = new SimStLogger(missController.getBrController());
 
@@ -3351,6 +3349,7 @@
      */
     public void startNewProblem() {
         
+    	System.out.println("#########NEW PROBLEM!!!!#####");
         killInteractiveLearningThreadIfAny();
         
         File wmeTypeFile = null;
@@ -6556,6 +6555,8 @@
     
     public String inquiryRuleActivation(String problemName, ProblemNode problemNode, String ruleName,  
             String selection, String action, String input, RuleActivationNode ran){
+    	
+    	//TODO THIS is the ugliest method I have ever seen. Please clean it up. -Maclellan
 
     	ProblemNode startNode = null;
     	String lastEquation = null;
@@ -6578,85 +6579,86 @@
         if(trace.getDebugCode("miss"))trace.out("miss", "           node: " + problemNode + "(" + lastEquation + ")");
         if(trace.getDebugCode("miss"))trace.out("miss", "      rule name: " + ruleName);
         if(trace.getDebugCode("miss"))trace.out("miss", "actualSelection: " + selection);
+        if(trace.getDebugCode("miss"))trace.out("miss", "   actualAction: " + action);
         if(trace.getDebugCode("miss"))trace.out("miss", "    actualInput: " + input);
         
-        String mtStatus =
-            this.isUseCacheOracleInquiry() ? 
-                    lookupRuleActivationStatus(problemName, lastEquation, "", selection, action, input) :
-                        null;
+        String mtStatus = null;
+        if (this.isUseCacheOracleInquiry()){
+                mtStatus = lookupRuleActivationStatus(problemName, lastEquation, "", selection, action, input);
+        }
 
-                    if (mtStatus == null) {
-                        // if there is no FALSE/FELAS  *or*  !isIlFromBrd, ask the respective oracle
-                        if (input.toUpperCase().indexOf("FALSE") == -1 && input.toUpperCase().indexOf("FELAS") == -1) {
+        if (mtStatus == null) {
+                // if there is no FALSE/FELAS  *or*  !isIlFromBrd, ask the respective oracle
+                if (input.toUpperCase().indexOf("FALSE") == -1 && input.toUpperCase().indexOf("FELAS") == -1) {
 
-                            if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_TS)) {
-                                mtStatus = inquiryRaTutoringService(selection, action, input, problemNode, problemName + ".brd") ;
-                            } 
-                            else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_HO)) {
-                                // ruleActivated.getName()
-                                mtStatus = inquiryRuleActivationOracle(selection, action, input, problemNode, problemName, ruleName, ran);
-                            }
-                            else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_CL)) {
-                                String problem = startNode.getName();
-                                mtStatus = inquiryClAlgebraTutor(selection, action, input, problemNode, problem);
-                            }
-                            else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_BRD)){
-                                // in interactive interactive learning, we pass current BRD (like here)
-                                // but in batch interactive learning, we want to pass whatever BRD is used at the time
-                                mtStatus = inquiryRuleActivationBRD(selection, action, input, problemNode.getName(), currentBrdPath);
-                            }
-                            else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_VOID)){
-                                mtStatus = "VOID";
-                            }
-                            else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_TUTOR_SOLVER)) {
-                                //String problem = getBrController().getProblemModel().getStartNode().getName();
-                                mtStatus = inquirySolverTutor(selection, action, input, problemNode, problemName);
-                            }
-                            else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_TUTOR_SOLVERV2)) {
+                    if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_TS)) {
+                        mtStatus = inquiryRaTutoringService(selection, action, input, problemNode, problemName + ".brd") ;
+                    } 
+                    else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_HO)) {
+                        // ruleActivated.getName()
+                        mtStatus = inquiryRuleActivationOracle(selection, action, input, problemNode, problemName, ruleName, ran);
+                    }
+                    else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_CL)) {
+                        String problem = startNode.getName();
+                        mtStatus = inquiryClAlgebraTutor(selection, action, input, problemNode, problem);
+                    }
+                    else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_BRD)){
+                        // in interactive interactive learning, we pass current BRD (like here)
+                        // but in batch interactive learning, we want to pass whatever BRD is used at the time
+                        mtStatus = inquiryRuleActivationBRD(selection, action, input, problemNode.getName(), currentBrdPath);
+                    }
+                    else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_VOID)){
+                        mtStatus = "VOID";
+                    }
+                    else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_TUTOR_SOLVER)) {
+                        //String problem = getBrController().getProblemModel().getStartNode().getName();
+                        mtStatus = inquirySolverTutor(selection, action, input, problemNode, problemName);
+                    }
+                    else if (getRuleActivationTestMethod().equalsIgnoreCase(RA_TEST_METHOD_TUTOR_SOLVERV2)) {
 
-                            	String problem = "";
-                            	if(getSsInteractiveLearning()== null || !getSsInteractiveLearning().isTakingQuiz())
-                            		problem = startNode.getName();
-                            	else 
-                            		problem = problemName; // When taking quiz use the problemName as we donot have reference to the BR_Controller
-                            	
-                                mtStatus = builtInInquiryClTutor(selection, action, input, problemNode, problem);
+                            String problem = "";
+                            if(getSsInteractiveLearning()== null || !getSsInteractiveLearning().isTakingQuiz())
+                                    problem = startNode.getName();
+                            else 
+                                    problem = problemName; // When taking quiz use the problemName as we donot have reference to the BR_Controller
+                            
+                        mtStatus = builtInInquiryClTutor(selection, action, input, problemNode, problem);
+                        
+                        if (getSsInteractiveLearning() != null && getSsInteractiveLearning().isTakingQuiz()){
+                            
+                            // If what the SimStudent did at this step is wrong ask from Oracle what would have been the right step
+                            if(mtStatus.equals(EdgeData.CLT_ERROR_ACTION)){
+                           
+                                AskHint hint = getCorrectSAI(getBrController(), problemNode);
                                 
-                                if (getSsInteractiveLearning() != null && getSsInteractiveLearning().isTakingQuiz()){
-                                    
-                                    // If what the SimStudent did at this step is wrong ask from Oracle what would have been the right step
-                                    if(mtStatus.equals(EdgeData.CLT_ERROR_ACTION)){
-                                   
-                                        AskHint hint = getCorrectSAI(getBrController(), problemNode);
-                                        
-                                        logger.simStLog(SimStLogger.SIM_STUDENT_QUIZ,SimStLogger.CONFIRMATION_REQUEST_CL_ACTION,
-                                                getProblemStepString(), mtStatus, "", new Sai(selection, action, input), mtStatus.equals(EdgeData.CORRECT_ACTION),
-                                                hint.getSelection(), hint.getAction(), hint.getInput());
-                                    } else {
-                                        logger.simStLog(SimStLogger.SIM_STUDENT_QUIZ,SimStLogger.CONFIRMATION_REQUEST_CL_ACTION,
-                                                getProblemStepString(), mtStatus, "", new Sai(selection, action, input), mtStatus.equals(EdgeData.CORRECT_ACTION),
-                                                "","","");
-                                    }
-                                }
+                                logger.simStLog(SimStLogger.SIM_STUDENT_QUIZ,SimStLogger.CONFIRMATION_REQUEST_CL_ACTION,
+                                        getProblemStepString(), mtStatus, "", new Sai(selection, action, input), mtStatus.equals(EdgeData.CORRECT_ACTION),
+                                        hint.getSelection(), hint.getAction(), hint.getInput());
+                            } else {
+                                logger.simStLog(SimStLogger.SIM_STUDENT_QUIZ,SimStLogger.CONFIRMATION_REQUEST_CL_ACTION,
+                                        getProblemStepString(), mtStatus, "", new Sai(selection, action, input), mtStatus.equals(EdgeData.CORRECT_ACTION),
+                                        "","","");
                             }
-                            // Sun Aug 20 16:26:02 LDT 2006 :: Noboru
-                            // Keep record of the oracles for rule activation status for the future use
-                            if (mtStatus != null && this.isUseCacheOracleInquiry()) {
-                                addRuleActivationStatus(problemName, lastEquation, "", selection, action, input, mtStatus);
-                                saveRuleActivationStatus();
-                            }
-                        } else {
-                            mtStatus = EdgeData.CLT_ERROR_ACTION;
                         }
                     }
-                    if(trace.getDebugCode("miss"))trace.out("miss", "inquiryRuleActivation: >>>>> result: " + mtStatus);
-                    
-                    return mtStatus;
-    }
-    
+                    // Sun Aug 20 16:26:02 LDT 2006 :: Noboru
+                    // Keep record of the oracles for rule activation status for the future use
+                    if (mtStatus != null && this.isUseCacheOracleInquiry()) {
+                        addRuleActivationStatus(problemName, lastEquation, "", selection, action, input, mtStatus);
+                        saveRuleActivationStatus();
+                    }
+                } else {
+                    mtStatus = EdgeData.CLT_ERROR_ACTION;
+                }
+        }
+        if(trace.getDebugCode("miss"))trace.out("miss", "inquiryRuleActivation: >>>>> result: " + mtStatus);
 
+        return mtStatus;
+}
+
+
+
     
-    
     /*public String calcProblemStepString()
     {
         ProblemNode currentNode = getBrController().getCurrentNode();
@@ -6843,7 +6845,7 @@
 	    	List wholeAgenda = ssRete.getAgendaAsList(null);
 	    	root.createChildren(wholeAgenda, false);
 	    	List children = root.getChildren();
-	    	JessModelTracing jmt = ssRete.getMt().getModelTracing();
+	    	JessModelTracing jmt = ssRete.getJessModelTracing();
 
 	    	for(int i=0; i< children.size(); i++) {
 
@@ -7431,7 +7433,6 @@
     // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
     // Asking to a human oracle
     // 
-
     public String inquiryRuleActivationOracle(String actualSelection, String actualAction, String actualInput, 
                                               ProblemNode node, String problemName, String ruleName,
                                               RuleActivationNode ran) {
@@ -7469,13 +7470,19 @@
       		  	long end = Calendar.getInstance().getTimeInMillis();
         	} 
 
+        	try{
         	hint = new AskHintInBuiltClAlgebraTutor(getBrController(), node);
+        	}
+        	catch(Exception ex){
+        		
         	
-        	if(logger.getLoggingEnabled())
-        	{
-	        	logger.simStLog(SimStLogger.SIM_STUDENT_INFO_RECEIVED, SimStLogger.CONFIRMATION_REQUEST_ACTION, step, 
-	        			"", title+":"+msg, sai,node,hint.getSelection(), hint.getAction(), hint.getInput(),0,msg);
         	}
+        	
+        //	if(logger.getLoggingEnabled())
+        	//{
+	       // 	logger.simStLog(SimStLogger.SIM_STUDENT_INFO_RECEIVED, SimStLogger.CONFIRMATION_REQUEST_ACTION, step, 
+	       // 			"", title+":"+msg, sai,node,hint.getSelection(), hint.getAction(), hint.getInput(),0,msg);
+        //	}
 
         	int oracle = displayConfirmMessage(title,message);
         	if (oracle == JOptionPane.YES_OPTION) {
@@ -7488,40 +7495,40 @@
         	// TODO: Need to figure out a unified way to handle all the interface actions at one place
         	// Model-tracing the student interface action (clicking either the Yes or No button) in response to the 
         	// SimStudent feedback request
-        	if(isSsMetaTutorMode()) {
+        	//if(isSsMetaTutorMode()) {
 
-	      		if(getBrController() != null && getBrController().getAmt() != null) {
-	      			if(oracle == JOptionPane.YES_OPTION) {
-	      				
-	      				// If the feedback is for done then run the model-tracer for done, ButtonPressed, -1
-	      				if(ruleName.replaceAll("MAIN::", "").contains(WorkingMemoryConstants.DONE_BUTTON_SELECTION)) {
-	      					getBrController().getAmt().handleInterfaceAction(WorkingMemoryConstants.DONE_BUTTON_SELECTION,
-	      							WorkingMemoryConstants.BUTTON_ACTION, WorkingMemoryConstants.BUTTON_INPUT);
-	      				} else {
-	      					getBrController().getAmt().handleInterfaceAction(WorkingMemoryConstants.YES_BUTTON_SELECTION,
-	      							WorkingMemoryConstants.BUTTON_ACTION, WorkingMemoryConstants.BUTTON_INPUT);
-	      				}
-	      			} else if(oracle == JOptionPane.NO_OPTION) {
-	      				
-	      				getBrController().getAmt().handleInterfaceAction(WorkingMemoryConstants.NO_BUTTON_SELECTION,
-	      						WorkingMemoryConstants.BUTTON_ACTION, WorkingMemoryConstants.BUTTON_INPUT);	      				
-	      			}
-	      		}
-        	}
+	      	//	if(getBrController() != null && getBrController().getAmt() != null) {
+	      	//		if(oracle == JOptionPane.YES_OPTION) {
+	      	//			
+	      	//			// If the feedback is for done then run the model-tracer for done, ButtonPressed, -1
+	      	//			if(ruleName.replaceAll("MAIN::", "").contains(WorkingMemoryConstants.DONE_BUTTON_SELECTION)) {
+	      	//				getBrController().getAmt().handleInterfaceAction(WorkingMemoryConstants.DONE_BUTTON_SELECTION,
+	      	//						WorkingMemoryConstants.BUTTON_ACTION, WorkingMemoryConstants.BUTTON_INPUT);
+	      	//			} else {
+	      	//				getBrController().getAmt().handleInterfaceAction(WorkingMemoryConstants.YES_BUTTON_SELECTION,
+	      	//						WorkingMemoryConstants.BUTTON_ACTION, WorkingMemoryConstants.BUTTON_INPUT);
+	      	//			}
+	      	//		} else if(oracle == JOptionPane.NO_OPTION) {
+	      	//			
+	      	//			getBrController().getAmt().handleInterfaceAction(WorkingMemoryConstants.NO_BUTTON_SELECTION,
+	      	//					WorkingMemoryConstants.BUTTON_ACTION, WorkingMemoryConstants.BUTTON_INPUT);	      				
+	      	//		}
+	      	//	}
+        	//}
         	
-        	if(getSsInteractiveLearning() != null)
-        	{
-        		//if(isSsMetaTutorMode()) {
-        		//	getModelTraceWM().getEventHistory().add(0, getModelTraceWM().new Event(SimStLogger.INPUT_VERIFY_ACTION));
-        		//}
-        		if(logger.getLoggingEnabled())
-        		{
-        			int verifyDuration = (int) (Calendar.getInstance().getTimeInMillis() - verifyRequestTime);
-	        		logger.simStLog(SimStLogger.SIM_STUDENT_INFO_RECEIVED, SimStLogger.INPUT_VERIFY_ACTION, step, 
-	        				status, title+":"+msg, sai,node,hint.getSelection(), hint.getAction(), hint.getInput(), verifyDuration,msg);
-        		}
-        		
-        	}
+        	//if(getSsInteractiveLearning() != null)
+        	//{
+        	//	//if(isSsMetaTutorMode()) {
+        	//	//	getModelTraceWM().getEventHistory().add(0, getModelTraceWM().new Event(SimStLogger.INPUT_VERIFY_ACTION));
+        	//	//}
+        	//	if(logger.getLoggingEnabled())
+        	//	{
+        	//		int verifyDuration = (int) (Calendar.getInstance().getTimeInMillis() - verifyRequestTime);
+	        //		logger.simStLog(SimStLogger.SIM_STUDENT_INFO_RECEIVED, SimStLogger.INPUT_VERIFY_ACTION, step, 
+	        //				status, title+":"+msg, sai,node,hint.getSelection(), hint.getAction(), hint.getInput(), verifyDuration,msg);
+        	//	}
+        	//	
+        	//}
         }
         
         if(trace.getDebugCode("miss"))trace.out("miss", "Oracle status ==> " + status );
@@ -8582,6 +8589,9 @@
     public /*private*/ boolean changeInstructionNameFor(String name, String oldSkillName, 
             ProblemNode node, Instruction instruction) {
 
+        if(trace.getDebugCode("rr"))
+            trace.out("rr", "Changing instruction name from " + oldSkillName + " to " + name);
+
         // Rename the skill name
         instruction.setName( name );
         // Relocate the instruction for now it has different name
Index: edu/cmu/pact/miss/SimStInteractiveLearning.java
===================================================================
--- edu/cmu/pact/miss/SimStInteractiveLearning.java	(revision 20342)
+++ edu/cmu/pact/miss/SimStInteractiveLearning.java	(working copy)
@@ -23,6 +23,7 @@
 import pact.CommWidgets.JCommTable;
 import pact.CommWidgets.JCommTable.TableCell;
 import pact.CommWidgets.JCommTable.TableExpressionCell;
+import pact.CommWidgets.JCommWidget;
 import edu.cmu.pact.BehaviorRecorder.Controller.BR_Controller;
 import edu.cmu.pact.BehaviorRecorder.Dialogs.LoadFileDialog;
 import edu.cmu.pact.BehaviorRecorder.ProblemModel.Graph.EdgeData;
@@ -33,6 +34,8 @@
 import edu.cmu.pact.BehaviorRecorder.ProblemModel.Graph.Groups.LinkGroup;
 import edu.cmu.pact.BehaviorRecorder.View.RuleLabel;
 import edu.cmu.pact.Utilities.trace;
+import edu.cmu.pact.ctat.MessageObject;
+import edu.cmu.pact.ctat.MsgType;
 import edu.cmu.pact.ctat.model.CtatModeModel;
 import edu.cmu.pact.jess.MTRete;
 import edu.cmu.pact.jess.RuleActivationNode;
@@ -1554,12 +1557,12 @@
 	        	simSt.stepDemonstrated(hint.node, hint.getSai(), hint.edge, null);
 	        	int foaCount = 0;
 	        	for(int i=0; i< simSt.numCurrentFoA(); i++) {
-	        		if((((TableExpressionCell) (simSt.getCurrentFoA().elementAt(i)).getWidget()).getText()).length() > 0)
-	        			++foaCount;
+	        		//if((((TableExpressionCell) (simSt.getCurrentFoA().elementAt(i)).getWidget()).getText()).length() > 0)
+	        				++foaCount;
 	        	}
 
 	        	// If the two FoA's have not been specified don't try to learn the step
-	        	if(foaCount == 2)
+	        	if(foaCount >= 1)
 	        	{
 	        		trace.out("ss", "Hint: "+hint+" Node: "+hint.node);
 	        		successful = simSt.changeInstructionName(hint.skillName, hint.node);
@@ -1609,7 +1612,7 @@
 	        		Instruction badInstr = simSt.lookupInstructionWithNode( hint.node );
 	        		simSt.deleteBadInstruction(badInstr);
 	        		
-		        	TableExpressionCell cell = null;
+		        	JCommWidget widget = null;
 		        	String selection = hint.getSai().getS();
 		        	
 		        	//Revert back to previous state
@@ -1634,13 +1637,15 @@
 		        	//If the problem has changed from the one we are working on, then don't do the display, kill learning
 		        	if(brController.getProblemModel() != null && problemName.equals(brController.getProblemModel().getProblemName()))
 		        	{
-		        		//put failed input in cell, editable, to remind student of what didn't work
-			        	if(brController.lookupWidgetByName(selection) instanceof TableExpressionCell)
-			        	{
-			        		cell = (TableExpressionCell)brController.lookupWidgetByName(selection);
-			        		cell.getTable().singleCellReset(selection);
-			        		cell.setText(hint.getSai().getI());
-			        	}
+		        		MessageObject mo = MessageObject.create(MsgType.INTERFACE_ACTION);
+		        		mo.setSelection(hint.getSelection());
+		        		mo.setAction(hint.getAction());
+		        		mo.setInput(hint.getInput());
+		        		
+		        		brController.handleCommMessage(mo);
+
+                        //widget = (JCommWidget)brController.lookupWidgetByName(selection);
+                        //widget.performAction(hint.getSai());
 			        	
 			        	hint.getRetry(brController, currentNode);
 		        	}
Index: edu/cmu/pact/miss/console/controller/MissController.java
===================================================================
--- edu/cmu/pact/miss/console/controller/MissController.java	(revision 20342)
+++ edu/cmu/pact/miss/console/controller/MissController.java	(working copy)
@@ -343,7 +343,7 @@
      *
      */
     public void startNewProblem() {
-	getSimSt().startNewProblem();
+        getSimSt().startNewProblem();
     }
 
     /**
@@ -1547,7 +1547,7 @@
 			wrapper = new SimStPLEWrapper((BR_Controller) controller);
 		else
 			wrapper = new TutorWrapper(controller);
-        if (simStPleOnArg)
+        if (ssContestArg)
         	wrapper = new GameShowWrapper((BR_Controller) controller);
         return wrapper;
 	}
Index: edu/cmu/pact/miss/jess/AMT.java
===================================================================
--- edu/cmu/pact/miss/jess/AMT.java	(revision 20342)
+++ edu/cmu/pact/miss/jess/AMT.java	(working copy)
@@ -43,7 +43,7 @@
 		
 		this.controller = controller;
 		ssRete = new SimStRete((BR_Controller)controller);
-		ssRete.setJmt(null);
+		//ssRete.setJmt(null);
 		aPlusModelTracing = new APlusModelTracing(ssRete, controller);
 		ssRete.setAmt(aPlusModelTracing);
 	}
Index: edu/cmu/pact/miss/jess/APlusModelTracing.java
===================================================================
--- edu/cmu/pact/miss/jess/APlusModelTracing.java	(revision 20342)
+++ edu/cmu/pact/miss/jess/APlusModelTracing.java	(working copy)
@@ -8,6 +8,8 @@
 import java.util.Queue;
 import java.util.Vector;
 
+import javax.swing.JOptionPane;
+
 import jess.Context;
 import jess.Fact;
 import jess.JessException;
@@ -308,6 +310,10 @@
 		studentInput = input;
 		nodeSeq.clear();
 		
+		
+		System.out.println("Start model tracing for selection:" + selection +"  -   action:" + action + "   -   input: "+ input);
+		
+		
 		try {
 			
 			//ssRete.clear(); // Added
@@ -378,6 +384,8 @@
 
 		iterativeDeepening(ssRete.getMaxDepth(), selection, action, input);
 		
+		System.out.println("nodeSeq:" + nodeSeq );
+		
 		if(nodeSeq.size() > 0) {
 			if(isHint || isCLHint)
 				setHintNode((RuleActivationNode)(nodeSeq.get(0)));
Index: edu/cmu/pact/miss/userDef/algebra/CopyTerm.java
===================================================================
--- edu/cmu/pact/miss/userDef/algebra/CopyTerm.java	(revision 20342)
+++ edu/cmu/pact/miss/userDef/algebra/CopyTerm.java	(working copy)
@@ -15,6 +15,7 @@
     public String apply(Vector args) 
     {
 	return (String)args.get(0);
+
     }
 
 }
Index: pact/CommWidgets/UniversalToolProxy.java
===================================================================
--- pact/CommWidgets/UniversalToolProxy.java	(revision 20342)
+++ pact/CommWidgets/UniversalToolProxy.java	(working copy)
@@ -23,6 +23,7 @@
 
 import javax.swing.JComponent;
 import javax.swing.JFrame;
+import javax.swing.JOptionPane;
 
 import org.jdom.Element;
 
